QMLFileType.js

```diff
@@ -0,0 +1,224 @@
+                var r = translated;
+                if (!translated || ( this.API.utils.cleanString(res.getSource()) !== this.API.utils.cleanString(r.getSource()) &&
+                    this.API.utils.cleanString(res.getSource()) !== this.API.utils.cleanString(r.getKey()))) {
+                    if (r) {
+                        logger.trace("extracted   source: " + this.API.utils.cleanString(res.getSource()));
+                        logger.trace("translation source: " + this.API.utils.cleanString(r.getSource()));
+                    }
+                    var note = r && 'The source string has changed. Please update the translation to match if necessary. Previous source: "' + r.getSource() + '"';
+                    var newres = res.clone();
+                    newres.setTargetLocale(locale);
+                    newres.setTarget((r && r.getTarget()) || res.getSource());
+                    newres.setState("new");
+                    newres.setComment(note);
+
+                    this.newres.add(newres);
+
+                    logger.trace("No translation for " + res.reskey + " to " + locale);
+                } else {
+                    if (res.reskey != r.reskey) {
+                        // if reskeys don't match, we matched on cleaned string.
+                        //so we need to overwrite reskey of the translated resource to match
+                        r = r.clone();
+                        r.reskey = res.reskey;
+                    }
+
+                    file = resFileType.getResourceFile(locale);
+                    file.addResource(r);
+                    logger.trace("Added " + r.reskey + " to " + file.pathName);
+                }
+            }.bind(this));
+        }.bind(this));
+    }
+
+    resources = this.pseudo.getAll().filter(function(resource) {
+        return resource.datatype === this.datatype;
+    }.bind(this));
+
+    for (var i = 0; i < resources.length; i++) {
+        res = resources[i];
+        if (res.getTargetLocale() !== this.project.sourceLocale && res.getSource() !== res.getTarget()) {
+            file = resFileType.getResourceFile(res.getTargetLocale());
+            file.addResource(res);
+            logger.trace("Added " + res.reskey + " to " + file.pathName);
+        }
+    }
+};
+
+QMLFileType.prototype.newFile = function(path) {
+    return new QMLFile({
+        project: this.project,
+        pathName: path,
+        type: this
+    });
+};
+
+QMLFileType.prototype.getDataType = function() {
+    return this.datatype;
+};
+
+QMLFileType.prototype.getResourceTypes = function() {
+    return {};
+};
+
+/**
+ * Return the name of the node module that implements the resource file type, or
+ * the path to a QML file that implements the resource filetype.
+ * @returns {Function|undefined} node module name or path, or undefined if this file type does not
+ * need resource files
+ */
+QMLFileType.prototype.getResourceFileType = function() {
+    //return JsonResourceFileType;
```

Comment:
Does QML not have resource file types?

Reply:
It should be `TSResourceFileType`. I will fix it.